File: src/main/java/org/trp/shincolle/entity/base/ShipFollowDecisionResolver.java
package org.trp.shincolle.entity.base;

final class ShipFollowDecisionResolver {
    private ShipFollowDecisionResolver() {
    }

    static boolean shouldTryTeleport(State state) {
        return ShipBrainRecoverySupport.shouldTryTeleportRecovery(
                state.recovery(),
                new ShipRecoveryDecisionResolver.State(
                        state.forceRecovery(),
                        state.ownerDistanceSqr(),
                        ShipAiNumbers.TELEPORT_DISTANCE_SQ
                ),
                ShipAiNumbers.FOLLOW_TELEPORT_COOLDOWN_TICKS
        );
    }

    record State(
            ShipMovementRecoveryState recovery,
            boolean forceRecovery,
            double ownerDistanceSqr
    ) {
    }
}

File: src/main/java/org/trp/shincolle/entity/base/ShipFollowDecisionResolver.kt
package org.trp.shincolle.entity.base

internal object ShipFollowDecisionResolver {
    fun shouldTryTeleport(state: State): Boolean {
        return ShipBrainRecoverySupport.shouldTryTeleportRecovery(
            state.recovery,
            ShipRecoveryDecisionResolver.State(
                state.forceRecovery,
                state.ownerDistanceSqr,
                ShipAiNumbers.TELEPORT_DISTANCE_SQ
            ),
            ShipAiNumbers.FOLLOW_TELEPORT_COOLDOWN_TICKS
        )
    }

    @JvmRecord
    internal data class State(
        val recovery: ShipMovementRecoveryState?,
        val forceRecovery: Boolean,
        val ownerDistanceSqr: Double
    )
}
